### PR TITLE
Additional type hints for config module.

### PR DESCRIPTION
**Pull request by clokep**
Adds some missing type hints to `synapse.config` on my journey to fully type it.

It will be several more PRs before that's done, unfortunately.

### CODE REVIEW

File: synapse/config/appservice.py [hunk 1]
```diff
@@ -30,12 +31,12 @@
 class AppServiceConfig(Config):
     section = "appservice"
 
-    def read_config(self, config, **kwargs):
```
Comment (DMRobertson) [inline synapse/config/appservice.py:33]: `config: Dict[str, Any]`?

kwargs seems to be `config_dir_path: str` and `data_dir_path: str` but I don't think these are frequently used. But let me not scope creep...

File: synapse/config/appservice.py [hunk 2]
```diff
@@ -30,12 +31,12 @@
...
+    def read_config(self, config, **kwargs) -> None:
         self.app_service_config_files = config.get("app_service_config_files", [])
         self.notify_appservices = config.get("notify_appservices", True)
         self.track_appservice_user_ips = config.get("track_appservice_user_ips", False)
 
-    def generate_config_section(cls, **kwargs):
+    def generate_config_section(cls, **kwargs) -> str:
```
Comment (DMRobertson) [inline synapse/config/appservice.py:39]: Do these methods ever do anything other than returning a string literal unconditionally?

I guess maybe one of them fills in a hostname or similar.

(I'm just wondering if this might be better off just being a static yaml file somewhere. Again, scope creep though.)

File: synapse/config/appservice.py [hunk 3]
```diff
@@ -50,7 +51,9 @@ def generate_config_section(cls, **kwargs):
         """
 
 
-def load_appservices(hostname, config_files):
+def load_appservices(
+    hostname: str, config_files: List[str]
+) -> List[ApplicationService]:
     """Returns a list of Application Services from the config files."""
     if not isinstance(config_files, list):
         logger.warning("Expected %s to be a list of AS config files.", config_files)
```
Comment (DMRobertson) [inline synapse/config/appservice.py:59]: Bit surprised this is  a warning rather than a refusal to startup!

File: synapse/config/cas.py [hunk 4]
```diff
@@ -28,7 +29,7 @@ class CasConfig(Config):
 
     section = "cas"
 
-    def read_config(self, config, **kwargs):
+    def read_config(self, config, **kwargs) -> None:
```
Comment (DMRobertson) [inline synapse/config/cas.py:32]: `config: Dict[str, Any]` while we're here.

File: synapse/config/cas.py [hunk 5]
```diff
@@ -51,7 +52,7 @@ def read_config(self, config, **kwargs):
             self.cas_displayname_attribute = None
             self.cas_required_attributes = []
 
-    def generate_config_section(self, config_dir_path, server_name, **kwargs):
+    def generate_config_section(self, config_dir_path, server_name, **kwargs) -> str:
```
Comment (DMRobertson) [inline synapse/config/cas.py:55]: `str` for both named args I think?

File: synapse/config/oidc.py [hunk 6]
```diff
@@ -36,7 +36,7 @@
 class OIDCConfig(Config):
     section = "oidc"
 
-    def read_config(self, config, **kwargs):
+    def read_config(self, config, **kwargs) -> None:
```
Comment (DMRobertson) [inline synapse/config/oidc.py:39]: `config: Dict[str, Any]` too

File: synapse/config/saml2.py [hunk 7]
```diff
@@ -33,7 +34,7 @@
 )
 
 
-def _dict_merge(merge_dict, into_dict):
+def _dict_merge(merge_dict: dict, into_dict: dict) -> None:
     """Do a deep merge of two dicts
 
     Recursively merges `merge_dict` into `into_dict`:
```
Comment (DMRobertson) [inline synapse/config/saml2.py:40]: Perhaps emphasise that this edits `into_dict` in-place?

File: synapse/config/saml2.py [hunk 8]
```diff
@@ -184,7 +185,7 @@ def read_config(self, config, **kwargs):
 
     def _default_saml_config_dict(
         self, required_attributes: set, optional_attributes: set
```
Comment (DMRobertson) [inline synapse/config/saml2.py:187]: `Set[str]` for both, I think?

File: synapse/config/sso.py [hunk 9]
```diff
@@ -49,7 +49,7 @@ class SSOConfig(Config):
 
     section = "sso"
 
-    def read_config(self, config, **kwargs):
+    def read_config(self, config, **kwargs) -> None:
```
Comment (DMRobertson) [inline synapse/config/sso.py:52]: `config: Dict[str, Any]`

File: synapse/config/cache.py [hunk 10]
```diff
@@ -162,7 +164,7 @@ def generate_config_section(self, **kwargs):
           #sync_response_cache_duration: 2m
         """
 
-    def read_config(self, config, **kwargs):
+    def read_config(self, config, **kwargs) -> None:
```
Comment (DMRobertson) [inline synapse/config/cache.py:167]: `config: Dict[str, Any]`

File: synapse/config/database.py [hunk 11]
```diff
@@ -119,7 +120,7 @@ def __init__(self, *args, **kwargs):
 
         self.databases = []
 
-    def read_config(self, config, **kwargs):
+    def read_config(self, config, **kwargs) -> None:
```
Comment (DMRobertson) [inline synapse/config/database.py:123]: `config: Dict[str, Any]`

File: synapse/config/database.py [hunk 12]
```diff
@@ -163,12 +164,12 @@ def read_config(self, config, **kwargs):
...
             self.set_databasepath(database_path)
 
-    def generate_config_section(self, data_dir_path, **kwargs):
+    def generate_config_section(self, data_dir_path, **kwargs) -> str:
         return DEFAULT_CONFIG % {
             "database_path": os.path.join(data_dir_path, "homeserver.db")
         }
 
-    def read_arguments(self, args):
+    def read_arguments(self, args: argparse.Namespace) -> None:
```
Comment (DMRobertson) [inline synapse/config/database.py:172]: Aside: I wish there was a way for mypy to know which attributes would be available on `args` at typechecktime.

File: synapse/config/logger.py [hunk 13]
```diff
@@ -142,13 +143,13 @@
 class LoggingConfig(Config):
     section = "logging"
 
-    def read_config(self, config, **kwargs):
+    def read_config(self, config, **kwargs) -> None:
```
Comment (DMRobertson) [inline synapse/config/logger.py:146]: `config: Dict[str, Any]`

File: synapse/config/server.py [hunk 14]
```diff
@@ -1291,7 +1294,9 @@ def is_threepid_reserved(reserved_threepids, threepid):
     return False
 
 
-def read_gc_thresholds(thresholds):
+def read_gc_thresholds(
+    thresholds: Optional[List[Any]],
```
Comment (DMRobertson) [inline synapse/config/server.py:1298]: Maybe `List[SupportsInt]` if you're feeling fancy!

Comment (DMRobertson) [reply synapse/config/server.py:1298]: Also, below on +1314, looks like `listener` could be `Dict: [str, Any]`

**Review by DMRobertson — COMMENTED** [review]
Thanks for this. A few thoughts:

File: synapse/config/appservice.py (hunk 1 above)
Comment (clokep) [reply synapse/config/appservice.py:33]: I have a separate branch built on this which adds type hints to `read_config`.

File: synapse/config/appservice.py (hunk 2 above)
Comment (clokep) [reply synapse/config/appservice.py:39]: Some of them return dynamic strings.

File: synapse/config/server.py (hunk 14 above)
Comment (clokep) [reply synapse/config/server.py:1298]: I'm not sure. That's what is *expected* by these methods, but not what is necessarily passed in?

File: synapse/config/database.py (hunk 12 above)
Comment (clokep) [reply synapse/config/database.py:172]: You mean if you could note what type of arguments exist in the `Namespace`?

File: synapse/config/appservice.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/config/appservice.py:33]: Ahh, my mistake.

File: synapse/config/database.py (hunk 12 above)
Comment (DMRobertson) [reply synapse/config/database.py:172]: I was thinking more that I'd like mypy to be able to deduce which args would exist, but an explict annotation would be good too!

**Comment by DMRobertson** [comment]
@clokep I think this needs a linter script run.

File: synapse/config/server.py (hunk 14 above)
Comment (clokep) [reply synapse/config/server.py:1298]: I'm not really sure what the best practice is here, but this is where we're taking any YAML junk and ensuring it is correct, so I think it would be a lie to say it has to accept integers?

**Comment by clokep** [comment]
> @clokep I think this needs a linter script run.

Doh, fixed!

File: synapse/config/appservice.py (hunk 1 above)
Comment (clokep) [reply synapse/config/appservice.py:33]: No problem! I should have included it in the description. :) There's a lot to these classes, so trying to break it up logically...but maybe only logically in my head. :)

File: synapse/config/server.py (hunk 14 above)
Comment (DMRobertson) [reply synapse/config/server.py:1298]: Agreed. (The argument I had in mind was that mypy could detect type errors if we described the intended inputs. But it's not going to be able to do that when the yaml comes at runtime.)

**Review by DMRobertson — APPROVED** [review]
Looks great. Bring on the next round!
